packages/labs/analyzer/src/test/browser/utils.ts

```diff
@@ -0,0 +1,215 @@
+
+export type InMemoryAnalyzerInit = {
+  typescript: TypeScript;
+  lang: Language;
+  files?: Files;
+};
+
+/**
+ * An Analyzer that analyzes an in-memory set of files passed into the
+ * constructor, which may be updated after the fact using the `setFile()` API.
+ */
+export class InMemoryAnalyzer extends Analyzer {
+  private _cache: Cache;
+  private _dirty = false;
+  private _lang: Language;
+
+  constructor({typescript, lang, files = {}}: InMemoryAnalyzerInit) {
+    const ts = typescript;
+    const cache: Cache = Object.fromEntries(
+      Object.entries(files).map(([name, content]) => [
+        normalize(name),
+        {content, version: 0},
+      ])
+    );
+    const host = createHost(typescript, cache, lang);
+    const service = ts.createLanguageService(host, ts.createDocumentRegistry());
+    let program: ts.Program;
+    super({
+      typescript,
+      getProgram: () => {
+        if (program === undefined || this._dirty) {
+          this._dirty = false;
+          program = service.getProgram()!;
+        }
+        return program;
+      },
+      fs: {
+        ...host,
+        useCaseSensitiveFileNames: false,
+      },
+      path,
+    });
+    this._cache = cache;
+    this._lang = lang;
+  }
+
+  setFile(name: string, content: string) {
+    const fileName = normalize(getSourceFilename(name, this._lang));
+    const prev = this._cache[fileName] ?? {version: -1};
+    this._cache[fileName] = {content, version: prev.version + 1};
+    this._dirty = true;
+  }
+}
+
+// Note these functions assume a specific setup in tsconfig.json for tests of TS
+// projects using these helpers, namely that the `rootDir` is 'src' and the
+// `outDir` is 'out'
+
+export const getSourceFilename = (f: string, lang: Language) =>
+  path.normalize(
+    lang === 'ts'
+      ? path.join(path.dirname(f), 'src', path.basename(f) + '.ts')
+      : f + '.js'
+  );
+
+export const getOutputFilename = (f: string, lang: Language) =>
+  path.normalize(
+    lang === 'ts'
+      ? path.join(path.dirname(f), 'out', path.basename(f) + '.js')
+      : path.normalize(f + '.js')
+  );
+
+/**
+ * Creates a ts.LanguageServiceHost that reads from an in-memory cache for all
+ * files except TS standard libs, which are read off of disk.
+ */
+const createHost = (typescript: TypeScript, cache: Cache, lang: Language) => {
+  console.log('createHost', lang);
+  return {
+    getScriptFileNames: () =>
+      Object.keys(cache).filter((s) => s.endsWith('.ts') || s.endsWith('.js')),
+    getScriptVersion: (fileName: string) =>
+      cache[normalize(fileName)].version.toString(),
+    getScriptSnapshot: (fileName: string) => {
+      if (isLib(fileName)) {
+        console.log('requested lib', fileName);
+        return undefined;
+        // return fs.existsSync(fileName)
+        //   ? ts.ScriptSnapshot.fromString(
+        //       fs.readFileSync(path.normalize(fileName), 'utf-8')
+        //     )
+        //   : undefined;
```

Comment:
I think this is for stuff like `lib.dom.d.ts`, so I guess this is another spot where (something) will need to pre-load the in-memory cache with this stuff too, same as the `deps`.